_source/_posts/2021-09-24-java-records.md

```diff
@@ -0,0 +1,556 @@
+
+A `record` is implicitly final, and cannot be abstract. You cannot enhance it later by extension, as the compiler will output the following error:
+
+```
+Cannot inherit from final 'com.okta.developer.records.EndOfGame'
+```
+
+A `record` cannot extend any class, not even its implicit superclass `Record`. But it can implement interfaces. Using the `extends` with records clause will cause the following error:
+
+```
+No extends clause allowed for record
+```
+
+A `record` will not have write accessors and the implicitly declared fields are final, and not modifiable via reflection. Moreover, a `record` cannot declare instance fields outside the `record` _header_. Records embody an _immutable by default_ policy, usually applied to data carrier classes.
+
+```
+Instance field is not allowed in record
+Cannot assign a value to final variable 'id'
+```
+
+Read accessors can be declared explicitly, but should never silently alter the Record state. **Review Record semantics before making explicit declarations of automatic members.**
+
+A record without any constructor declaration will be given a canonical constructor with the same signature as the header, which will assign all the private fields from the arguments. The canonical constructor can be declared explicitly in the standard syntax, or in a compact form:
+
+```java
+public EndOfGame {
+    Objects.requireNonNull(id);
+    Objects.requireNonNull(date);
+    Objects.requireNonNull(timeOfDay);
+    Objects.requireNonNull(mentalState);
+    Objects.requireNonNull(damageTaken);
+    Objects.requireNonNull(damageToPlayers);
+    Objects.requireNonNull(damageToStructures);
+}
+```
+
+In the compact form, the parameters are implicit, the private fields cannot be assigned inside the body and are automatically assigned **at the end**. Its purpose is to allow focusing on validation, making defensive copies of mutable values, or some other value processing.
+
+## Serialization, Encoding, Mapping
+
+Record instances can extend `Serializable`, but the serialization and deserialization processes cannot be customized. Serialization and deserialization methods like `writeObject`,  `readObject` can be implemented, but will be ignored.
+
+As `spring-web` module provides Jackson JSON encoders and decoders, and Java Record support was added to Jackson in release 2.12, records can be used for REST API request and response mapping.
+
+Records cannot be used with JPA/Hiberante. Spring Data modules that do not use the object mapping of the underlying data store (like JPA) do support Record, as persistence construction detection works as with other classes. For Spring Data MongoDB, the general recommendation is to _stick to immutable objects_, because they are straightforward to materialize by calling the constructor. It is also recommended to provide an all-args constructor, allowing to skip property population for optimal performance. Java Record semantics align with these guidelines.
```

Comment:
```suggestion
Records cannot be used with JPA/Hibernate. Spring Data modules that do not use the object mapping of the underlying data store (like JPA) do support Record, as persistence construction detection works as with other classes. For Spring Data MongoDB, the general recommendation is to _stick to immutable objects_, because they are straightforward to materialize by calling the constructor. It is also recommended to provide an all-args constructor, allowing to skip property population for optimal performance. Java Record semantics align with these guidelines.
```

Reply:
Made bold and added a brief detail.